go/parquet/metadata/column_chunk.go

```diff
@@ -0,0 +1,385 @@
+	if hasDict {
+		thriftEncodings = append(thriftEncodings, format.Encoding(c.props.DictionaryIndexEncoding()))
+		if c.props.Version() == parquet.V1 {
+			thriftEncodings = append(thriftEncodings, format.Encoding_PLAIN)
+		} else {
+			thriftEncodings = append(thriftEncodings, format.Encoding(c.props.DictionaryPageEncoding()))
+		}
+	} else { // no dictionary
+		thriftEncodings = append(thriftEncodings, format.Encoding(c.props.EncodingFor(c.column.Path())))
+	}
+
+	thriftEncodings = append(thriftEncodings, format.Encoding(parquet.Encodings.RLE))
+	// Only PLAIN encoding is supported for fallback in V1
+	// TODO(zeroshade): Use user specified encoding for V2
+	if dictFallback {
+		thriftEncodings = append(thriftEncodings, format.Encoding_PLAIN)
+	}
+	c.chunk.MetaData.Encodings = thriftEncodings
+
+	thriftEncodingStats := make([]*format.PageEncodingStats, 0, len(dictEncodingStats)+len(dataEncodingStats))
+	for k, v := range dictEncodingStats {
+		thriftEncodingStats = append(thriftEncodingStats, &format.PageEncodingStats{
+			PageType: format.PageType_DICTIONARY_PAGE,
+			Encoding: format.Encoding(k),
+			Count:    v,
+		})
+	}
+	for k, v := range dataEncodingStats {
+		thriftEncodingStats = append(thriftEncodingStats, &format.PageEncodingStats{
+			PageType: format.PageType_DATA_PAGE,
+			Encoding: format.Encoding(k),
+			Count:    v,
+		})
+	}
+	c.chunk.MetaData.EncodingStats = thriftEncodingStats
+
+	encryptProps := c.props.ColumnEncryptionProperties(c.column.Path())
+	if encryptProps != nil && encryptProps.IsEncrypted() {
+		ccmd := format.NewColumnCryptoMetaData()
+		if encryptProps.IsEncryptedWithFooterKey() {
+			ccmd.ENCRYPTION_WITH_FOOTER_KEY = format.NewEncryptionWithFooterKey()
+		} else {
+			ccmd.ENCRYPTION_WITH_COLUMN_KEY = &format.EncryptionWithColumnKey{
+				KeyMetadata:  []byte(encryptProps.KeyMetadata()),
+				PathInSchema: c.column.ColumnPath(),
+			}
+		}
+		c.chunk.CryptoMetadata = ccmd
+
+		encryptedFooter := c.props.FileEncryptionProperties().EncryptedFooter()
+		encryptMetadata := !encryptedFooter || !encryptProps.IsEncryptedWithFooterKey()
+		if encryptMetadata {
+			// Serialize and encrypt ColumnMetadata separately
+			// Thrift-serialize the ColumnMetaData structure,
+			// encrypt it with the column key, and write to encrypted_column_metadata
+			serializer := thrift.NewThriftSerializer()
+			data, err := serializer.Write(context.Background(), c.chunk.MetaData)
+			if err != nil {
+				return err
+			}
+			var buf bytes.Buffer
+			metaEncryptor.Encrypt(&buf, data)
+			c.chunk.EncryptedColumnMetadata = buf.Bytes()
+
+			if encryptedFooter {
+				c.compressedSize = c.chunk.MetaData.GetTotalCompressedSize()
+				c.chunk.MetaData = nil
+			} else {
+				// Keep redacted metadata version for old readers
+				c.chunk.MetaData.Statistics = nil
+				c.chunk.MetaData.EncodingStats = nil
+			}
+		}
+	}
+	return nil
+}
+
+// WriteTo will always return 0 as the int64 since the thrift writer library
```

Comment:
Hmm, I wonder if we should keep track of the bytes or explicit return a -1. Is the president for this elsewhere?

Reply:
So, in internal/utils/write_utils.go I do have what i called a `TellWrapper` which wrapped an io.Writer in order to keep track of the bytes written so that it could add a `Tell()` function to tell the current position which I use elsewhere. 

The main reason why I didn't do something like that here was that I didn't want every call to `WriteTo` to create an object wrapping the io.Writer in order to count the bytes written. It's just unfortunate that the current thrift interface for a `TStruct` has a `Write` function but doesn't return the number of bytes it wrote.

I don't remember there being anywhere else that I do this so you might be right that it makes sense to explicitly return a -1 here instead of a 0. Personally as long as the comment documents what is being returned, I don't much mind whether it's a 0 or a -1 as long as we're explicit about it.